dali/pipeline/data/dynamic_tensor_view.h

```diff
@@ -0,0 +1,623 @@
+  explicit DynamicTensorView(const TensorView<Backend, const T, ndim> &other) = delete;
+  template <typename T, int other_ndim>
+  explicit DynamicTensorView(const TensorView<Backend, const T, other_ndim> &other) = delete;
+  // @}
+
+  /**
+   * @name Convert to strongly-typed TensorView
+   *
+   * Requires match between runtime and static types, no-op if DataType = DynamicType
+   */
+  // @{
+  template <typename DataType, int other_ndim = ndim>
+  std::enable_if_t<!std::is_same<std::remove_const_t<DataType>, DynamicType>::value,
+                   TensorView<Backend, DataType, other_ndim>>
+  to_static_type() const & {
+    DALI_ENFORCE(type_id == TypeTable::GetTypeId<DataType>(),
+                 make_string("Calling type does not match view data type, requested type: ",
+                             TypeTable::GetTypeId<DataType>(), " current view type: ", type_id));
+    detail::check_compatible_ndim<ndim, other_ndim>();
+    return {static_cast<DataType *>(data), shape};
+  }
+
+  template <typename DataType, int other_ndim = ndim>
+  std::enable_if_t<!std::is_same<std::remove_const_t<DataType>, DynamicType>::value,
+                   TensorView<Backend, DataType, other_ndim>>
+  to_static_type() && {
+    DALI_ENFORCE(type_id == TypeTable::GetTypeId<DataType>(),
+                 make_string("Calling type does not match view data type, requested type: ",
+                             TypeTable::GetTypeId<DataType>(), " current view type: ", type_id));
+    detail::check_compatible_ndim<ndim, other_ndim>();
+    return {static_cast<DataType *>(data), std::move(shape)};
+  }
+
+  template <typename DataType, int other_ndim = ndim>
+  std::enable_if_t<std::is_same<std::remove_const_t<DataType>, DynamicType>::value,
+                   TensorView<Backend, DataType, other_ndim>>
+  to_static_type() const & {
+    detail::check_compatible_ndim<ndim, other_ndim>();
+    return {data, shape, type_id};
+  }
+
+  template <typename DataType, int other_ndim = ndim>
+  std::enable_if_t<std::is_same<std::remove_const_t<DataType>, DynamicType>::value,
+                   TensorView<Backend, DataType, other_ndim>>
+  to_static_type() && {
+    detail::check_compatible_ndim<ndim, other_ndim>();
+    return {data, std::move(shape), type_id};
+  }
+  // @}
+
+  /**
+   * @brief Change the ndim, compatible with TensorView API
+   */
+  template <int other_ndim>
+  DynamicTensorView<Backend, other_ndim> to_static() const {
+    static_assert(other_ndim != DynamicDimensions,
+                  "Conversion to static only allowed for static shape");
+    static_assert(ndim == other_ndim || ndim == DynamicDimensions, "Cannot convert to other ndim");
+    return {data, shape.template to_static<other_ndim>(), type_id};
+  }
+
+  using Base::data;
+  using Base::shape;
+  using Base::type_id;
+};
+
+template <typename Backend, int ndim = DynamicDimensions>
+struct ConstDynamicTensorView : DynamicTensorViewBase<Backend, const void, ndim> {
```

Comment:
Seeing this copy-paste here makes me reconsider using `void`. Alternatively, how about `const DynamicType`?